File: leetcode/Remove_Nth_Node_From_End_of_List.java
import java.util.*;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode[] a = new ListNode[100];
        int x = 0;
        ListNode cur = head;
        while (cur != null) {
            a[x] = cur;
            x++;
            cur = cur.next;
        }
        int y = x - n;
        if (y == 0) head = head.next;
        else a[y - 1].next = a[y].next;
        return head;
    }
}

File: leetcode/Remove_Nth_Node_From_End_of_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        stack<ListNode *> reverse_order;
        ListNode *cur = head;
        while (cur != NULL) {
            reverse_order.push(cur);
            cur = cur->next;
        }
        int reverse_index = 1;
        while (reverse_index < n) {
            reverse_order.pop();
            ++reverse_index;
        }
        ListNode *deleting = reverse_order.top();
        reverse_order.pop();
        if (!reverse_order.empty()) {
            ListNode *pre = reverse_order.top();
            reverse_order.pop();
            pre->next = deleting->next;
        } else {
            head = deleting->next;
        }
        delete deleting;
        return head;
    }
};
